tests/providers/conftest.py

```diff
@@ -1,20 +1,95 @@
+from exchange.providers import Provider, AzureProvider, OllamaProvider, OpenAiProvider
+from exchange.providers.ollama import OLLAMA_MODEL
+
+ProviderList = List[Tuple[Type[Provider], str]]
+
+# provider configuration for all tests except vision
+all_providers: ProviderList = [
+    (OllamaProvider, os.getenv("OLLAMA_MODEL", OLLAMA_MODEL)),
+    (OpenAiProvider, os.getenv("OPENAI_MODEL", "gpt-4o-mini")),
+    (AzureProvider, os.getenv("AZURE_MODEL", "gpt-4o-mini")),
+]
+
+
+def mark_parametrized(providers: ProviderList = None, expected_params: Dict[Type[Provider], Tuple[Any]] = None):
+    """When expected_params is present, we assume it is a VCR test which needs
+    to fake ENV variables to avoid validation failures initializing the
+    provider. This is done via the provider_cls function"""
+
+    if providers is None:
+        providers = all_providers
+    # Create ids based on provider class names in lowercase without 'Provider' suffix
+    ids = [cls.__name__.replace("Provider", "").lower() for cls, _ in providers]
+
+    # When there are no parameters, we assume this is an integration test. So,
+    # all we do is parameterize the tests for each input provider.
+    if expected_params is None:
+
+        def decorator(test: Callable):
+            return pytest.mark.parametrize("provider_cls,model", providers, ids=ids)(test)
+
+        return decorator
+
+    # Parameters means this is a VCR test, as we can't test for explicit (due
+    # to models not being 100pct deterministic wrt the specific tokens
+    # generated or count of them).
+    #
+    # When running as a VCR test, we need to parameterize the tests with values
+    # specific to a provider, and also fake the ENV variables the provider uses
+    # so that it doesn't fail on initialization.
+
+    # provider_cls is what we use to fake the ENV variables, which is a fixture
+    # and cannot be passed as a function. We have to pass it by name instead.
+    provider_cls_name = provider_cls.__name__
+
+    def decorator(test: Callable):
+        # This wraps the real test function after the provider_cls parameter is
+        # initialized (with `provider_cls`). Notably, this chooses parameters
+        # based on the provider, and passes them after provider_cls, model.
+        def add_params(provider_cls: Type[Provider], model: str) -> Tuple[Any]:
+            expected_values = expected_params.get(provider_cls)
+            if expected_values is None:
+                raise ValueError(f"No expectations found for provider class: {provider_cls}")
+            return test(provider_cls, model, *expected_values)
+
+        # This part parameterizes the test for each provider, initializing
+        # `provider_cls` with fake ENV (via `indirect`). Each test includes
+        # provider-specific parameters which are expanded in `add_params`.
+        return pytest.mark.parametrize("provider_cls,model", providers, ids=ids, indirect=[provider_cls_name])(
```

Comment:
not sure how to document this, but there seems to be a requirement for the "indirect" target to be initialized with a function exactly matching the parameter name. This was a bit confusing to me, but might be normal for folks familiar with pytest. Anyway, mentioning it as it was at least new to me.